### PR TITLE
Move Spark-indpendent Table debug to cudf Java

### PR DESCRIPTION
**Pull request by gerashegalov**
## Description
spark-rapids has [code for debugging JNI Tables/Columns][1] that is useful for debugging during dev work in cudf

This PR proposes to start moving it to cudf/java. spark-rapids will be updated to call into the cudf in a follow-up PR.

[1]: https://github.com/NVIDIA/spark-rapids/blob/b5cf25eef347d845bd77077d5cb9035262281f98/sql-plugin/src/main/java/com/nvidia/spark/rapids/GpuColumnVector.java

## Sample Usage with JShell
```Bash
(rapids) rapids@compose:~/cudf/java$ mvn dependency:build-classpath -Dmdep.outputFile=cudf-java-cp.txt
(rapids) rapids@compose:~/cudf/java$ jshell --class-path target/cudf-23.10.0-SNAPSHOT-cuda12.jar:$(< cudf-java-cp.txt) \
    -R -Dai.rapids.cudf.debug.output=log_error
```
```Java
|  Welcome to JShell -- Version 11.0.20
|  For an introduction type: /help intro

jshell> import ai.rapids.cudf.*;

jshell> Table tbl = new Table.TestBuilder().column(1,2,3,4,5,6).build()
tbl ==> Table{columns=[ColumnVector{rows=6, type=INT32, n ... e=140381937458144, rows=6}

jshell> TableDebug.get().debug("gera", tbl)
[main] ERROR ai.rapids.cudf.TableDebug - DEBUG gera Table{columns=[ColumnVector{rows=6, type=INT32, nullCount=Optional[0], offHeap=(ID: 4 7fad371d1a30)}], cudfTable=140381937458144, rows=6}
[main] ERROR ai.rapids.cudf.TableDebug - GPU COLUMN 0 - NC: 0 DATA: DeviceMemoryBufferView{address=0x7fad3be00000, length=24, id=-1} VAL: null
[main] ERROR ai.rapids.cudf.TableDebug - COLUMN 0 - INT32
[main] ERROR ai.rapids.cudf.TableDebug - 0 1
[main] ERROR ai.rapids.cudf.TableDebug - 1 2
[main] ERROR ai.rapids.cudf.TableDebug - 2 3
[main] ERROR ai.rapids.cudf.TableDebug - 3 4
[main] ERROR ai.rapids.cudf.TableDebug - 4 5
[main] ERROR ai.rapids.cudf.TableDebug - 5 6
```

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

**Review by revans2 — APPROVED** [review]
This looks fine. Just curious if we want to adjust/improve anything while we have the patient on the table?

Like passing in an output stream instead of always using stderr? Not that it is really needed.

**Comment by gerashegalov** [comment]
> Just curious if we want to adjust/improve anything while we have the patient on the table?

I am open to these and further suggestions

File: java/src/main/java/ai/rapids/cudf/TableDebug.java [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+  public static TableDebug get() {
+    return DEFAULT_DEBUG;
+  }
+
+  private final PrintStream out;
+  private TableDebug(PrintStream printStream) {
+    out = printStream;
+  }
+  /**
+   * Print to standard error the contents of a table. Note that this should never be
```
Comment (revans2) [inline java/src/main/java/ai/rapids/cudf/TableDebug.java:77]: nit: it is not necessarily "standard error" any more.

**Review by revans2 — APPROVED** [review]

File: java/src/main/java/ai/rapids/cudf/TableDebug.java [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+ * limitations under the License.
+ */
+package ai.rapids.cudf;
+
+import java.io.PrintStream;
+import java.util.Locale;
+
+public class TableDebug {
+
+
```
Comment (ttnghia) [inline java/src/main/java/ai/rapids/cudf/TableDebug.java:23]:
```suggestion

```

File: java/src/main/java/ai/rapids/cudf/TableDebug.java [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+  private static final String OUTPUT_STREAM = "ai.rapids.cudf.debug.output";
+  enum Output {
```
Comment (ttnghia) [inline java/src/main/java/ai/rapids/cudf/TableDebug.java:25]:
```suggestion

    enum Output {
```

File: java/src/main/java/ai/rapids/cudf/TableDebug.java [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+
+  public static TableDebug get() {
+    return DEFAULT_DEBUG;
+  }
+
+  private final PrintStream out;
+  private TableDebug(PrintStream printStream) {
+    out = printStream;
+  }
+  /**
```
Comment (ttnghia) [inline java/src/main/java/ai/rapids/cudf/TableDebug.java:76]:
```suggestion

  /**
```

File: java/src/main/java/ai/rapids/cudf/TableDebug.java [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+        }
+      }
+    } else if (col.getType() == DType.LIST) {
+      try (ColumnView child = col.getChildColumnView(0)) {
+        debugGPUAddrs(name + ":DATA", child);
+      }
+    }
+  }
+
+
```
Comment (ttnghia) [inline java/src/main/java/ai/rapids/cudf/TableDebug.java:125]:
```suggestion

```

File: java/src/main/java/ai/rapids/cudf/TableDebug.java [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+      debug(String.valueOf(col), table.getColumn(col));
+    }
+  }
+
+  /**
+   * Print to standard error the contents of a column. Note that this should never be
+   * called from production code, as it is very slow.  Also note that this is not production
+   * code.  You might need/want to update how the data shows up or add in support for more
+   * types as this really is just for debugging.
+   * @param name the name of the column to print out.
```
Comment (ttnghia) [inline java/src/main/java/ai/rapids/cudf/TableDebug.java:96]:
```suggestion

   * @param name the name of the column to print out.
```

File: java/src/main/java/ai/rapids/cudf/TableDebug.java [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+  private final PrintStream out;
+  private TableDebug(PrintStream printStream) {
+    out = printStream;
+  }
+  /**
+   * Print to standard error the contents of a table. Note that this should never be
+   * called from production code, as it is very slow.  Also note that this is not production
+   * code.  You might need/want to update how the data shows up or add in support for more
+   * types as this really is just for debugging.
+   * @param name the name of the table to print out.
```
Comment (ttnghia) [inline java/src/main/java/ai/rapids/cudf/TableDebug.java:81]:
```suggestion

   * @param name the name of the table to print out.
```

**Review by ttnghia — APPROVED** [review]
There are many nit formatting issues. Please check carefully before merging.

**Comment by gerashegalov** [comment]
> There are many nit formatting issues. Please check carefully before merging.

I am going to hit reformat with ` dev/cudf_java_styles.xml`

**Pull request merged**